### PR TITLE
Transaction Analyser in Kotlin

### PR DESCRIPTION
**Pull request by cjxuan1995**

### CODE REVIEW

File: src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionAnalyserAppTest.kt [hunk 1]
```diff
@@ -0,0 +1,41 @@
+package com.codingchallenge.transactionanalyser
```
Comment (caladev) [inline src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionAnalyserAppTest.kt:1]: You mentioned on Slack that you added tests after. 
I want you to start getting in the mindset that tests drive the design of your code - that is, the practice of TDD. Test Driven Development. It is not an answer to everything I know, but in my view it will change your view on how you bring new features in and have tests that cover them. 
Many years ago Guy asked me to view the following video by the man himself uncle bob: https://vimeo.com/43734265 yes he is awkward but pay attention to what he is trying to teach and what it achieves by writing tests before your implementation. 
Also see: https://blog.cleancoder.com/uncle-bob/2014/12/17/TheCyclesOfTDD.html
I want you to try this when we are pairing together next.

File: src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import com.codingchallenge.transactionanalyser.domain.TransactionBuilder
+import java.io.BufferedReader
+import java.io.InputStream
+import java.io.InputStreamReader
+import java.util.stream.Collectors
+
+class TransactionLoader(private val input: InputStream) {
+    fun loadTransactionsFromCsv(): List<Transaction>{
+        val transactions: List<Transaction>
+        BufferedReader(InputStreamReader(input)).use {
```
Comment (caladev) [inline src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt:15]: have not seen us use `.use` in the team much. this is a great use case and I like the brevity.

File: src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            transactions = it.lines()
+                .skip(1)
+                .map(this::buildTransactionFromCsvLine)
+                .collect(Collectors.toList())
+        }
+        return transactions
+    }
+
+    private fun buildTransactionFromCsvLine(line: String): Transaction{
```
Comment (caladev) [inline src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt]: is there a better way we could have thought about passing the values from the CSV through? 
Here we are using magic numbers https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad 
Be careful with this, and think about how the data might scale, if new columns are added then we would have to change every number after that column change.
Could we perhaps have used the kotlin-csv library and mapped the headings of the CSV itself, and used those names to map to the builder here rather than with 0,1,2,3 etc? or any other way?

File: src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionProcessor.kt [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package com.codingchallenge.transactionanalyser
+
+import com.codingchallenge.transactionanalyser.domain.Transaction
+import com.codingchallenge.transactionanalyser.domain.UserInput
+import java.math.BigDecimal
+
+class TransactionProcessor {
+    fun filterTransactionsByUserInput(transactions: List<Transaction>, input: UserInput): List<Transaction>{
+        return transactions
```
Comment (caladev) [inline src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionProcessor.kt:9]: consider the use of comments here. Generally I feel comments are making your code less readable here.
Your code itself is doing many things in each line so perhaps you thought a comment will help the reader? think again and consider the person reading the code and why that might be. We can refactor our code further an avoid the comments completely. Think about how each `.filter` could be read as plain english. 

For example: let's look at line 11.
I could make a smaller private function below this class (or a Kotlin extension function even) which extracts out `it.fromAccountId == input.accountId` and we call that function `isMatchingFromAccount(...)` or something and then we extract out `it.toAccountId == input.accountId` to another small private function or a Kotlin extension function and call it `isMatchingToAccountId(...)` or something, we then have made small functions that do one thing and one thing well. 
Your line 11 will then read as something like:
`.filter { isMatchingToAccountId(...) || isMatchingFromAccount(...) }`
it makes it clearer for the code to express itself in plain english rather than in a comment

File: src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionProcessor.kt [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+            // find the transactions related to the account
+            .filter { it.fromAccountId == input.accountId || it.toAccountId == input.accountId }
+            // find the transactions where the createdAt is after the "from" time and before the "to" time
+            .filter { it.createdAt > input.from && it.createdAt < input.to }
+            // Remove transactions that get reversed
+            .filter { ft -> transactions.none { t -> ft.transactionId == t.relatedTransaction } }
```
Comment (caladev) [inline src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionProcessor.kt]: rather than using `ft` consider naming it to what it actually is. Naming it `ft` makes the reader guess.
Same with `t`. 
We shouldn't have to read code for the first time and have this big cognitive load to figure out what something is doing and why. Think about the pseudocode in your head of what it's doing and see how closely you can use the langauge to bring that out.

File: src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionAnalyserApp.kt [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package com.codingchallenge.transactionanalyser
+
+import java.lang.IllegalArgumentException
+import java.time.format.DateTimeFormatter
+
+class TransactionAnalyserApp {
+    fun runApp(){
+        val transactionLoader = TransactionLoader(
+            this::class.java.classLoader.getResourceAsStream("transactions.csv") ?: throw IllegalArgumentException("File not found")
```
Comment (caladev) [inline src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionAnalyserApp.kt:9]: Just an FYI.
Ensure you are familiar with the Exceptions best practices in kotlin. They are different to in Java. Obviously Java has checked exceptions and unchecked. Kotlin does not have checked exceptions.

https://elizarov.medium.com/kotlin-and-exceptions-8062f589d07 Roman Elizarov is on the Kotlin team and used to be the tech lead of the language team itself. His post here is very insightful on where we should be considering exception handling in our Kotlin code.
Think about code you have seen where there are `try`/`catch` blocks polluting an entire chain of functions. And consider what he is saying here to keep that in one place, and let the exception bubble up gracefully.

Comment (caladev) [reply src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionAnalyserApp.kt:9]: It will be wise to follow Roman Elizarov on Twitter as well

File: src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
+package com.codingchallenge.transactionanalyser
+
+import com.codingchallenge.transactionanalyser.domain.Transaction
+import com.codingchallenge.transactionanalyser.domain.TransactionBuilder
+import java.io.BufferedReader
+import java.io.InputStream
+import java.io.InputStreamReader
+import java.util.stream.Collectors
+
+class TransactionLoader(private val input: InputStream) {
```
Comment (caladev) [inline src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt:12]: I hope you can see here a good example of the Kotlin language itself being **immutability** as part of its design. It is a big thing we should always be thinking of, and consider whether you are writing Java style code in Kotlin, or are you truly writing immutable Kotlin code in Kotlin. (or are you just writing Java style to be compatible with Java libraries, hehe).

What I mean by this specifically is that your `TransactionLoader` is taking in a value in its constructor and it is a `val`. We aren't making getters and setters where `input` could be modified at any point later. It gives us more confidence in `TransactionLoader` to have a predictable output from what you pass in. Of course that is not always the case but notice the language is encouraging us to work this way.

File: src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    fun loadTransactionsFromCsv(): List<Transaction>{
+        val transactions: List<Transaction>
+        BufferedReader(InputStreamReader(input)).use {
+            transactions = it.lines()
+                .skip(1)
```
Comment (caladev) [inline src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt]: here you are doing `skip(1)` 
Again come back to my previous comment about cognitive load on a new reader of this code.
I know you skipped 1 because it's the heading row. But think about someone guessing that or trying to see you are doing that without knowing why you had to. 
Could we extract that into a separate function, perhaps, call it `skipHeaderRow()` which would make it more obvious to the reader? Maybe not that easy on the InputStreamReader but something to just have a think about.

File: src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+    private fun buildTransactionFromCsvLine(line: String): Transaction{
+        val values = line.split(",").map { it.trim() }
+        var transactionBuilder = TransactionBuilder(values[0])
+            .withFromAccountId(values[1])
+            .withToAccountId(values[2])
+            .withCreatedAt(values[3])
+            .withAmount(values[4])
+            .withType(values[5])
+        if (values.size == 7){
```
Comment (caladev) [inline src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt]: you are checking the size of the number of columns to determine if the transaction is a reversal. 
What if I added more columns with different data? 
Bring your code design closer to the data itself so you don't have to introduce this "7 fields equals a reversal" rule that you have essentially created here. What I mean by that is you introduced a concept, a rule, something that is unique to this codebase but not if the data structure changes. The CSV file **has** a field with the label `REVERSAL`. That is the way you know it's a reversal from the data itself. Perhaps we could change our code to use _that_ as the only rule that it is a reversal, not if it has a certain number of columns.

Be adaptable to change and think about the rules you are introducing yourself that is different from the 'rules of the data' in this case.

File: src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        assertEquals(transactions[1].amount.compareTo(BigDecimal.valueOf(10.50)), 0)
+        assertEquals(transactions[1].transactionType, TransactionType.PAYMENT)
+        assertNull(transactions[1].relatedTransaction)
+
+        assertEquals(transactions[3].transactionId, "TX10004")
+        assertEquals(transactions[3].fromAccountId, "ACC334455")
+        assertEquals(transactions[3].toAccountId, "ACC998877")
+        assertEquals(transactions[3].createdAt, LocalDateTime.parse("20/10/2018 19:45:00", formatter))
+        assertEquals(transactions[3].amount.compareTo(BigDecimal.valueOf(10.50)), 0)
+        assertEquals(transactions[3].transactionType, TransactionType.REVERSAL)
```
Comment (caladev) [inline src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt]: there are different schools of thought on this, but generally I believe one test should test one thing.
If something in the implementation changes, your test are there to cover you. Every time you make a small change to the implementation, you run the tests, they pass, you are ok. But if one thing changes in our implementation and the test fails and that test is doing **soooo** many things, we almost then need to go digging through the test itself to figure out where in the test it broke as well.
Consider small tests that have a single assert statement (or not more than one or two, fewer is better than lots). If a change is made to your implementation and a test breaks, that test should be very clear on what exactly went wrong. Don't make more work for yourself just because it's easier to add lots of assert statements in one place; treat the Clean Code principles equal in your test code as in your implementation code. It will serve you well.

Comment (caladev) [reply src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt]: In this case we have 3 different transactions being tested all at once. At the very least, we could assert differently for those. 
In addition, perhaps we could have one assert statement on a single transaction object, rather than for every field one by one on that transaction

File: src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+import java.time.format.DateTimeFormatter
+
+internal class TransactionLoaderTest{
+    private val transactionLoader = TransactionLoader(
+        this::class.java.classLoader.getResourceAsStream("transactions.csv") ?: throw IllegalArgumentException("File not found")
+    )
+    private val formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss")
+
+    @Test
+    fun givenCsv_WhenLoadTransactionsFromCsv_ThenTheExpectedTransactionsReturned(){
```
Comment (caladev) [inline src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt]: it is generally good practice in kotlin tests to use the special Junit feature of writing the function name without camel case and use backticks. For example, `fun givenCsv_WhenLoadTransactions...`  can turn into `fun ``givenCsv_WhenLoadTransactions```

Comment (caladev) [reply src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt]: sorry I can't get the formatting right. What I mean is like this https://stackoverflow.com/a/65923081/11281137

Comment (caladev) [inline src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt]: You will also see in many of our real prod code tests, we are big fans of using `@Nested` tests. so you could have one layer of given / when / then in a class hierarchy where a child class can run in a nested way but all tests still run.
Good example here worth reading: https://kotlintesting.com/nested-tests-in-junit5/

File: src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        assertEquals(transactions[3].transactionId, "TX10004")
+        assertEquals(transactions[3].fromAccountId, "ACC334455")
+        assertEquals(transactions[3].toAccountId, "ACC998877")
+        assertEquals(transactions[3].createdAt, LocalDateTime.parse("20/10/2018 19:45:00", formatter))
+        assertEquals(transactions[3].amount.compareTo(BigDecimal.valueOf(10.50)), 0)
+        assertEquals(transactions[3].transactionType, TransactionType.REVERSAL)
+        assertEquals(
+            transactions[3].relatedTransaction,
+            "TX10002"
+        ) // REVERSAL transaction should have non-null relatedTransaction
```
Comment (caladev) [inline src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt]: coming back to comments, it is the same princple I described in your `TransactionProcessor`. You didn't really need a comment here if your code can express what it's doing without that comment. 
This could have been separated into its own test which the test function name was called perhaps fun `reversal transaction should have non-null related transaction`() and your assert could have been in it. Rather than a comment.
Again consider clean code principles in tests themselves. Test code is just as important as implementation code.

**Comment by caladev** [comment]
I want you to checkout Guy's code, pull it down and review in your IDE - https://github.com/annoymousGiraf/banking-classification
I want you to see the use of kotlin coroutines and how they are used to perform requests in parallel for this same problem statement. Refer to the functions you see with `async` and `await`. We can discuss in person as well.
I am partway through https://www.amazon.com.au/Learning-Concurrency-Kotlin-efficient-applications-ebook/dp/B0791HNZ4J and I think it will be a good thing for you to read at some point.

File: src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt (hunk 11 above)
Comment (caladev) [reply src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt]: Edit: that is a kotlin feature not a Junit feature sorry

File: src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionAnalyserApp.kt (hunk 6 above)
Comment (cjxuan1995) [reply src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionAnalyserApp.kt:9]: > Just an FYI.
> Ensure you are familiar with the Exceptions best practices in kotlin. They are different to in Java. Obviously Java has checked exceptions and unchecked. Kotlin does not have checked exceptions.
> 
> https://elizarov.medium.com/kotlin-and-exceptions-8062f589d07 Roman Elizarov is on the Kotlin team and used to be the tech lead of the language team itself. His post here is very insightful on where we should be considering exception handling in our Kotlin code.
> Think about code you have seen where there are `try`/`catch` blocks polluting an entire chain of functions. And consider what he is saying here to keep that in one place, and let the exception bubble up gracefully.

Will have a closer look at how Kotlin handles exceptions.

File: src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt (hunk 7 above)
Comment (cjxuan1995) [reply src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt:12]: Seems like Kotlin has the `final` keyword baked into `val`.

File: src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt (hunk 3 above)
Comment (cjxuan1995) [reply src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt]: Yeah, magic numbers are bad. I didn't use external CSV libraries for the simplicity sake. Now looking back it seems to be a mistake(i.e. the code is not production ready). I think there are libraries out there that can handle this nicely (i.e. annotation on the class which maps fields to csv columns). Will have a look at what kotlin-csv offers.

File: src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionProcessor.kt (hunk 4 above)
Comment (cjxuan1995) [reply src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionProcessor.kt:9]: Very valid point. Readable code > comments for sure.

File: src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionAnalyserAppTest.kt (hunk 1 above)
Comment (cjxuan1995) [reply src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionAnalyserAppTest.kt:1]: Will definitely look into TDD.

File: src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt (hunk 11 above)
Comment (cjxuan1995) [reply src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt]: That's handy.

File: src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt (hunk 8 above)
Comment (cjxuan1995) [reply src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt]: Yeah, should be addressed by using an external csv library.

File: src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt (hunk 9 above)
Comment (cjxuan1995) [reply src/main/kotlin/com/codingchallenge/transactionanalyser/TransactionLoader.kt]: Yeah, I can see that it's a bad idea.

File: src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt (hunk 10 above)
Comment (cjxuan1995) [reply src/test/kotlin/com/codingchallenge/transactionanalyser/TransactionLoaderTest.kt]: Agree it should be broken down into smaller test methods. I was being lazy.
